tool/src/org/antlr/v5/automata/ATNOptimizer.java

```diff
@@ -99,25 +99,13 @@ private static void optimizeSets(Grammar g, ATN atn) {
 						throw new UnsupportedOperationException("Not yet implemented.");
 					}
 					IntervalSet set =  matchTransition.label();
-					List<Interval> intervals = set.getIntervals();
-					int n = intervals.size();
-					for (int k = 0; k < n; k++) {
-						Interval setInterval = intervals.get(k);
-						int a = setInterval.a;
-						int b = setInterval.b;
-						if (a != -1 && b != -1) {
-							for (int v = a; v <= b; v++) {
-								if (matchSet.contains(v)) {
-									// TODO: Token is missing (i.e. position in source is not displayed).
-									g.tool.errMgr.grammarError(ErrorType.CHARACTERS_COLLISION_IN_SET, g.fileName,
-											null,
-											CharSupport.getANTLRCharLiteralForChar(v),
-											CharSupport.getIntervalSetEscapedString(matchSet));
-									break;
-								}
-							}
-						}
+					IntervalSet intersection = matchSet.and(set);
```

Comment:
set1.and(set2) doesn't cater for interval.a == -1, so this could be a breaking change ?

Reply:
(not sure which scenario can produce `interval.a = -1` or `interval.b = -1`)

Reply:
> set1.and(set2) doesn't cater for interval.a == -1, so this could be a breaking change ?

It looks like these checks were introduced for correct iterating using `for` loop. But in my change `for` loop is not used and `matchSet.and(set)` correctly handles `-1` as well.

> (not sure which scenario can produce interval.a = -1 or interval.b = -1)

I agree, it's not very clear code. To figure out when it's possible, I was forced to run testsuite. I figured out it can be `-1` only for `EOF` symbol. See [`testLexerEOFInSet`](https://github.com/antlr/antlr5/blob/dev/tool-testsuite/test/org/antlr/v5/test/tool/TestATNDeserialization.java#L108), [`testEofClosure`](https://github.com/antlr/antlr5/blob/dev/tool-testsuite/test/org/antlr/v5/test/tool/TestToolSyntaxErrors.java#L362) as examples. But it's not a problem since `matchSet.and(set)` returns intersection and it correctly handles `EOF` as well.